### PR TITLE
AtomicCount: add arch guards and fallback to compiler intrinsics (#206)

### PR DESCRIPTION
**Pull request by neomantra**
The current implementation of `atomic_count` has unguarded inline
x86 assembly, causing failure on arm #206.

This commit adds preprocessor guards on the x86 assembly and falls back
to compiler intrinsic `__atomic_fetch_add` for other architectures.

----

Compared to previous attempts on this issue, the change is scoped to just the assembly language portion of the code.  It is minimal and there is zero change to platforms that work (x86).   

I broke out `arm` as its own `#elif` block as that's the platform that was failing and in case somebody with ARM assembly skills wants to fill that in.    Happy to change it to one `#else` if that is preferred.

### CODE REVIEW

**Comment by VivienCabannes** [comment]
It would be nice to approve this push request and to update the pipy repository in order to allow for `pip install quickfix` to work on aarch and arm architecture.

**Comment by realdanurbano** [comment]
Second FWIW. As I mention in a separate user's bug report; upgrading to Monterey 12.2, and subsequent upgrade to `CommandLineTools` fixed my note above WRT `CommandLineTools`.

**Review by wrc-marek — APPROVED** [review]

**Comment by shane-lamb** [comment]
bumping, would love for this to be merged as it's critical for the increasing numbers of M1 users these days

**Comment by jimmcslim** [comment]
I've checked out the fork of this PR and attempted to build but still getting errors on an M1 Mac, Monterey 12.3.

````
../../../src/C++/AtomicCount.h:165:27: warning: value size does not match register size specified by the constraint and modifier [-Wasm-operand-widths]
          "+m"(*pw), "=r"(r) : // outputs (%0, %1)
                          ^
../../../src/C++/AtomicCount.h:164:17: note: use constraint modifier "w"
          "xadd %1, %0":
                ^~
                %w1
../../../src/C++/AtomicCount.h:163:11: error: unrecognized instruction mnemonic
          "lock\n\t"
          ^
<inline asm>:1:2: note: instantiated into assembly here
        lock
        ^
In file included from FieldMapTestCase.cpp:28:
In file included from ../../../src/C++/FieldMap.h:30:
In file included from ../../../src/C++/MessageSorters.h:30:
In file included from ../../../src/C++/SharedArray.h:26:
../../../src/C++/AtomicCount.h:163:18: error: unrecognized instruction mnemonic, did you mean: add, addp, adds, addv, fadd, madd?
          "lock\n\t"
                 ^
<inline asm>:2:2: note: instantiated into assembly here
        xadd x9, [x8]
        ^
In file included from FieldMapTestCase.cpp:28:
In file included from ../../../src/C++/FieldMap.h:30:
In file included from ../../../src/C++/MessageSorters.h:30:
In file included from ../../../src/C++/SharedArray.h:26:
../../../src/C++/AtomicCount.h:163:11: error: unrecognized instruction mnemonic
          "lock\n\t"
          ^
<inline asm>:1:2: note: instantiated into assembly here
        lock
        ^
In file included from FieldMapTestCase.cpp:28:
In file included from ../../../src/C++/FieldMap.h:30:
In file included from ../../../src/C++/MessageSorters.h:30:
In file included from ../../../src/C++/SharedArray.h:26:
../../../src/C++/AtomicCount.h:163:18: error: unrecognized instruction mnemonic, did you mean: add, addp, adds, addv, fadd, madd?
          "lock\n\t"
                 ^
                 ````

**Review by pan93412 — APPROVED** [review]

**Comment by JimAmuro** [comment]
another m1 guy blocked by compilation failure... is there any gentleman with write access to the repo helping us merge it?
Excuse me, Mike @mgatny could you help? thanks in advance!

**Comment by neomantra** [comment]
To people following this, I just rebased this on the latest `master`, which includes some SSL functionality.   

I tested it on my M1 laptop; I don't have an M2.  For anybody following along, you may need to [disable AirPlay Receiver](https://stackoverflow.com/questions/4421633/who-is-listening-on-a-given-tcp-port-on-mac-os-x) to get `make check` Unit Tests to work.

**Comment by orenmnero** [comment]
#430 Gets rid of all this non-sense and brings in std::atomic

**Comment by orenmnero** [comment]
#430 switches to using std::atomic

**Pull request closed without merging**
